File: src/main/java/phoenix/utils/block/AnonimBlock.java
package phoenix.utils.block;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.material.Material;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.IBlockReader;
import phoenix.Phoenix;

import java.util.function.Supplier;

public class AnonimBlock extends Block
{
    public static Supplier<Block> create(Block.Properties properties)
    {
        return () ->new AnonimBlock(properties);
    }

    public static Supplier<Block> create(Material material)
    {
        return () ->new AnonimBlock(Properties.create(material));
    }
    public static Supplier<Block> create(Material material, Class<TileEntity> entityClass)
    {
        return () ->new AnonimBlock(Properties.create(material), entityClass);
    }

    public static Supplier<Block> create(Block.Properties properties, Class<TileEntity> classIn)
    {
        return () -> new AnonimBlock(properties, classIn);
    }
    Class<TileEntity> tile = null;
    public AnonimBlock(Block.Properties properties)
    {
        super(properties);
    }
    public AnonimBlock(Block.Properties properties, Class<TileEntity> classIn)
    {
        super(properties);
        tile = classIn;
    }

    @Override public boolean hasTileEntity(BlockState state){ return tile != null;	}


    @Override
    public TileEntity createTileEntity(BlockState state, IBlockReader world)
    {
        try
        {
            return tile.newInstance();
        }
        catch (Exception e)
        {
            Phoenix.LOGGER.error("Can not init tile: " + tile);
            return null;
        }
    }
}

File: src/main/java/phoenix/utils/block/AnonimBlock.kt
package phoenix.utils.block

import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.material.Material
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.IBlockReader
import net.minecraft.world.storage.loot.LootContext
import phoenix.Phoenix
import phoenix.utils.SizedArrayList
import java.util.function.Supplier


class AnonimBlock(properties: Properties) : Block(properties), ICustomGroup
{
    var tile: Class<TileEntity>? = null
    var group : ItemGroup = Phoenix.ASH
    constructor(properties: Properties, classIn: Class<TileEntity>) : this(properties)
    {
        tile = classIn
    }

    constructor(properties: Properties, classIn: Class<TileEntity>, itemGroup: ItemGroup) : this(properties)
    {
        tile = classIn
        group = itemGroup
    }

    constructor(properties: Properties, itemGroup: ItemGroup) : this(properties)
    {
        group = itemGroup
    }

    override fun hasTileEntity(state: BlockState): Boolean
    {
        return tile != null
    }

    override fun createTileEntity(state: BlockState, world: IBlockReader): TileEntity?
    {
        return try
        {
            tile!!.newInstance()
        }
        catch (e: Exception)
        {
            Phoenix.LOGGER.error("Can not init tile: $tile")
            null
        }
    }

    override fun getDrops(state: BlockState, builder: LootContext.Builder) = SizedArrayList.of(ItemStack(this))

    override fun getTab() = group

    companion object
    {
        fun create(properties: Properties)                               = Supplier { AnonimBlock(properties) }
        fun create(material: Material)                                   = Supplier { AnonimBlock(Properties.create(material)) }
        fun create(material: Material, entityClass: Class<TileEntity>)   = Supplier { AnonimBlock(Properties.create(material), entityClass) }
        fun create(properties: Properties, classIn: Class<TileEntity>)   = Supplier { AnonimBlock(properties, classIn) }
        fun create(material: Material, entityClass: Class<TileEntity>, itemGroup: ItemGroup)   = Supplier { AnonimBlock(Properties.create(material), entityClass, itemGroup) }
        fun create(properties: Properties, itemGroup: ItemGroup)   = Supplier { AnonimBlock(properties, itemGroup) }
    }
}
